doc/resiliency.md

```diff
@@ -0,0 +1,175 @@
+# Writing resilient code
+
+## Thread-safety considerations
+
+JSON-RPC is an inherently asynchronous protocol. Multiple concurrent requests are allowed.
+StreamJsonRpc will dispatch method calls as the requests are processed, even while prior requests are still running.
+`JsonRpc` always invokes local methods by posting to its `SynchronizationContext`.
+
+The default `SynchronizationContext` schedules work to the thread pool, which allows your code
+to respond to multiple requests concurrently. You should write your objects to be thread-safe to avoid
+malfunctions when clients call multiple requests at once by following thread-safe patterns.
+
+Making code thread-safe in the face of concurrency is the ideal solution, as it maximizes performance.
+When your code is not thread-safe, you should work to avoid data corruption and other malfunctions by taking
+either of the approaches described below to avoid concurrent execution.
+
+### Throttling concurrency to 1
+
+If your code has no particular thread affinity, but you want to prevent any concurrent execution,
+you may enclose each invokable method inside a semaphore. Be sure to wait for the semaphore asynchronously
+to avoid flooding your server's thread pool with waiters. Below is an example of a server target object that
+defends against concurrency as a thread-safety mitigation:
+
+```cs
+class Server
+{
+    AsyncSemaphore semaphore = new AsyncSemaphore(1);
+
+    public async Task DoSomething(CancellationToken cancellationToken)
+    {
+        using (semaphore.EnterAsync(cancellationToken))
```

Comment:
Wouldn't you need to await this?

Reply:
doh! Yes.

Reply:
Incidentally, my mistake would compile just fine, since `Task` derives from `IDisposable`. It's such a nefarious mistake that it was discussed on an analyzers forum a while back and I wrote [VSTHRD107](https://github.com/Microsoft/vs-threading/blob/master/doc/analyzers/VSTHRD107.md#vsthrd107-await-task-within-using-expression) to produce an error when it happens. Sadly, my analyzer doesn't run on markdown. 😬